validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java

```diff
@@ -38,39 +39,102 @@ public class BeaconProposerPreparer implements ValidatorTimingChannel {
   private static final Logger LOG = LogManager.getLogger();
 
   private final ValidatorApiChannel validatorApiChannel;
-  private final ValidatorIndexProvider validatorIndexProvider;
+  private Optional<ValidatorIndexProvider> validatorIndexProvider;
   private final ProposerConfigProvider proposerConfigProvider;
   private final Spec spec;
   private final Optional<Eth1Address> defaultFeeRecipient;
   private boolean firstCallDone = false;
 
-  public BeaconProposerPreparer(
+  private final Map<Integer, BeaconPreparableProposer> runtimeProposerMap =
+      new ConcurrentHashMap<>();
+  private final Map<Integer, BeaconPreparableProposer> configuredProposerMap =
+      new ConcurrentHashMap<>();
+
+  BeaconProposerPreparer(
       ValidatorApiChannel validatorApiChannel,
       ValidatorIndexProvider validatorIndexProvider,
       ProposerConfigProvider proposerConfigProvider,
       Optional<Eth1Address> defaultFeeRecipient,
       Spec spec) {
+    this(
+        validatorApiChannel,
+        Optional.of(validatorIndexProvider),
+        proposerConfigProvider,
+        defaultFeeRecipient,
+        spec);
+  }
+
+  public BeaconProposerPreparer(
+      ValidatorApiChannel validatorApiChannel,
+      Optional<ValidatorIndexProvider> validatorIndexProvider,
+      ProposerConfigProvider proposerConfigProvider,
+      Optional<Eth1Address> defaultFeeRecipient,
+      Spec spec) {
     this.validatorApiChannel = validatorApiChannel;
     this.validatorIndexProvider = validatorIndexProvider;
     this.proposerConfigProvider = proposerConfigProvider;
     this.defaultFeeRecipient = defaultFeeRecipient;
     this.spec = spec;
   }
 
+  public void initialize(final Optional<ValidatorIndexProvider> provider) {
+    this.validatorIndexProvider = provider;
+  }
+
   @Override
   public void onSlot(UInt64 slot) {
+    if (validatorIndexProvider.isEmpty()) {
+      return;
+    }
     if (slot.mod(spec.getSlotsPerEpoch(slot)).isZero() || !firstCallDone) {
       firstCallDone = true;
-
       sendPreparableProposerList();
     }
   }
 
+  public Optional<Eth1Address> getFeeRecipient(final BLSPublicKey publicKey) {
+    if (validatorIndexProvider.isEmpty()) {
+      return Optional.empty();
+    }
+    final Optional<Integer> validatorIndex =
+        validatorIndexProvider.orElseThrow().getValidatorIndex(publicKey);
+    return validatorIndex.flatMap(this::getFeeRecipientByValidatorIndex);
+  }
```

Comment:
hmm, I think we need to support setting and getting a fee recipient even before the validator index is known. You should be able to configure your validator client appropriately before making the deposit rather than having to wait for the validator to be active.

That probably means the config always has to be indexed by public key, not validator index. Not sure what the knock on effects of that will be though.